python/tvm/build_module.py

```diff
@@ -52,17 +57,58 @@ def __getattr__(self, name):
             return BuildConfig.defaults[name]
         return self._attr[name]
 
+    def decorator(self, func):
+        ''' decorate the pass function'''
+        def dump(*args, **kwargs):
+            '''dump function'''
+            retv = func(*args, **kwargs)
+            # pylint: disable=unidiomatic-typecheck
+            if type(retv) not in self._types:
+                return retv
+            pname = str(self._pass_id) + "_" + func.func_name + "_ir.cc"
+            with open(pname, "a") as f:
+                out = retv.body if isinstance(retv, container.LoweredFunc) else retv
+                print(out, file=f)
+                if isinstance(retv, container.Array):
+                    for x in retv:
+                        out = x.body if isinstance(x, container.LoweredFunc) else x
+                        print("---------", x.name, "\n", out, "\n-----------\n", file=f)
+                self._pass_id += 1
+            return retv
+        return dump
+
+    def decorate_irpass(self):
+        for k, v in vars(ir_pass).items():
+            vars(ir_pass)[k] = self.decorator(v) if isinstance(v, types.FunctionType) else v
+        schedule.ScheduleOps = self.decorator(schedule.ScheduleOps)
+        self._types = [v for k, v in vars(_stmt).items()]
+        self._types.append(container.Array)
+        self._types.append(container.LoweredFunc)
+
+    def decorate_custompass(self, add_lower_pass):
+        if self.dump_pass_ir is False:
+            return add_lower_pass
+        self._pass_id = 0
+        pass_list = [(x[0], self.decorator(x[1])) for x in add_lower_pass]
+        return pass_list
+
     def __enter__(self):
         # pylint: disable=protected-access
         self._old_scope = BuildConfig.current
         attr = BuildConfig.current._attr.copy()
         attr.update(self._attr)
         self._attr = attr
         BuildConfig.current = self
+        if self.dump_pass_ir is True:
+            self.decorate_irpass()
         return self
 
     def __exit__(self, ptype, value, trace):
         assert self._old_scope
+        if self.dump_pass_ir is True:
+            reload(ir_pass)
```

Comment:
reload is not safe when you have multiple-nest of decoration. 

with decorate1():
     with decorate2():
          both in effect
      decorate 1 in effect

Reply:
reload is not safe for nested scope, add recovery closure function to a stack and execute these recovery function in exit

```python
with decorate1():
    with decorate2():
        both in effect
     decorate1 in effect
```

Reply:
if decorates ir_pass module, we may need to handle nested case, if use temporary dict to manage functions in ir_pass, we don't need to do it.